File: node_second_line/src/runtime.mjs

import { digestData, loadJson, PRIMITIVE_IDENTITY } from './common.mjs';
import { evaluateRequest } from './profile_eval.mjs';
import { makePermit, makeRefusal } from './refusal.mjs';
import { attachWitnessCertificate } from './witness_certificate.mjs';
import { verifyBundleNode } from './verifier.mjs';

export class NodeSecondLineRunner {
  constructor(bundleDir) {
    this.bundleDir = bundleDir;
    this.verification = verifyBundleNode(bundleDir);
    this.bundleManifest = loadJson(`${bundleDir}/MANIFEST.sha256.json`);
    this.bundleDigest = digestData(this.bundleManifest);
    this.profile = loadJson(`${bundleDir}/PROFILE_SOURCE.json`);
  }
  validateRequest(request) {
    if (!request || typeof request !== 'object' || Array.isArray(request)) return 'REFUSE_SCHEMA_INVALID';
    const encoded = Buffer.from(JSON.stringify(request));
    if (encoded.length > 64 * 1024) return 'REFUSE_REQUEST_OVERSIZE';
    if (!request.object || typeof request.object !== 'object' || Array.isArray(request.object)) return 'REFUSE_SCHEMA_INVALID';
    return null;
  }
  execute(request, action) {
    const request_digest = digestData(request);
    if (!this.verification.overall_ok) {
      const out = makeRefusal({ code: this.verification.refusal_code || 'REFUSE_BUNDLE_VERIFICATION_FAILED', layer: 'bundle', bundle_digest: this.verification.bundle_digest, request_digest, verification_result: 'NOT_VERIFIED', refusal_reasons: this.verification.refusal_reasons || ['REFUSE_BUNDLE_VERIFICATION_FAILED'], details: { source: 'node-second-line.execute' } });
      out.overall_ok = false; return attachWitnessCertificate(out);
    }
    const requestProblem = this.validateRequest(request);
    if (requestProblem) {
      const out = makeRefusal({ code: requestProblem, layer: 'runtime', bundle_digest: this.bundleDigest, request_digest, refusal_reasons: [requestProblem], details: { source: 'node-second-line.execute' } });
      out.overall_ok = false; return attachWitnessCertificate(out);
    }
    if (!action || typeof action !== 'object' || Array.isArray(action) || JSON.stringify(Object.keys(action).sort()) !== JSON.stringify(['action_id','kind','payload'])) {
      const out = makeRefusal({ code: 'REFUSE_ACTION_DESCRIPTOR_INVALID', layer: 'action', bundle_digest: this.bundleDigest, request_digest, action_digest: null, refusal_reasons: ['REFUSE_ACTION_DESCRIPTOR_INVALID'], details: { source: 'node-second-line.execute' } });
      out.overall_ok = false; return attachWitnessCertificate(out);
    }
    const result = evaluateRequest(this.profile, request);
    if (result.verdict !== 'ALLOW') {
      const out = makeRefusal({ code: 'REFUSE_VERDICT_NOT_ALLOW', layer: 'action', bundle_digest: this.bundleDigest, request_digest, action_digest: digestData(action), refusal_reasons: ['REFUSE_VERDICT_NOT_ALLOW'], details: { source: 'node-second-line.execute', decisive_rule_ids: result.decisive_rule_ids, matched: result.matched, failed: result.failed, fail_closed: result.fail_closed } });
      out.overall_ok = false; return attachWitnessCertificate(out);
    }
    return attachWitnessCertificate(makePermit({ bundle_digest: this.bundleDigest, request_digest, action, decisive_rule_ids: result.decisive_rule_ids }));
  }
}

export function replayVerdictNode(bundleDir, request) {
  const runner = new NodeSecondLineRunner(bundleDir);
  const request_digest = digestData(request);
  if (!runner.verification.overall_ok) {
    const out = makeRefusal({ code: runner.verification.refusal_code || 'REFUSE_BUNDLE_VERIFICATION_FAILED', layer: 'bundle', bundle_digest: runner.verification.bundle_digest, request_digest, verification_result: 'NOT_VERIFIED', refusal_reasons: runner.verification.refusal_reasons || ['REFUSE_BUNDLE_VERIFICATION_FAILED'], details: { source: 'node-second-line.replay' } });
    out.overall_ok = false; return attachWitnessCertificate(out);
  }
  const requestProblem = runner.validateRequest(request);
  if (requestProblem) {
    const out = makeRefusal({ code: requestProblem, layer: 'runtime', bundle_digest: runner.bundleDigest, request_digest, refusal_reasons: [requestProblem], details: { source: 'node-second-line.replay' } });
    out.overall_ok = false; return attachWitnessCertificate(out);
  }
  const result = evaluateRequest(runner.profile, request);
  if (result.verdict !== 'ALLOW') {
    const out = makeRefusal({ code: 'REFUSE_VERDICT_NOT_ALLOW', layer: 'runtime', bundle_digest: runner.bundleDigest, request_digest, refusal_reasons: ['REFUSE_VERDICT_NOT_ALLOW'], details: { source: 'node-second-line.replay', decisive_rule_ids: result.decisive_rule_ids, matched: result.matched, failed: result.failed, fail_closed: result.fail_closed } });
    out.overall_ok = false; return attachWitnessCertificate(out);
  }
  return attachWitnessCertificate({
    primitive_identity: PRIMITIVE_IDENTITY,
    overall_ok: true,
    bundle_digest: runner.bundleDigest,
    request_digest,
    verdict: 'ALLOW',
    replay_binding: { primitive_identity: PRIMITIVE_IDENTITY, bundle_digest: runner.bundleDigest, request_digest },
    decisive_rule_ids: [...result.decisive_rule_ids],
    matched: [...result.matched],
    failed: [...result.failed],
    fail_closed: result.fail_closed,
  });
}

File: externalized_second_line/src/bcf_primitive_externalized/runtime.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from bcf_primitive_verifier.witness_certificate import attach_witness_certificate

from .common import digest_data, load_json
from .profile_eval import evaluate_request
from .refusal import make_permit, make_refusal
from .verifier import verify_bundle_externalized


class ExternalizedRunner:
    def __init__(self, bundle_dir: str | Path):
        self.bundle = Path(bundle_dir)
        self.verification = verify_bundle_externalized(self.bundle)
        self.manifest = load_json(self.bundle / 'MANIFEST.sha256.json')
        self.bundle_digest = digest_data(self.manifest)
        self.profile = load_json(self.bundle / 'PROFILE_SOURCE.json')

    def _validate_request(self, request: Any) -> str | None:
        if not isinstance(request, dict):
            return 'REFUSE_SCHEMA_INVALID'
        encoded = json.dumps(request, sort_keys=True, separators=(',', ':'), ensure_ascii=False).encode('utf-8')
        if len(encoded) > 64 * 1024:
            return 'REFUSE_REQUEST_OVERSIZE'
        if 'object' not in request or not isinstance(request['object'], dict):
            return 'REFUSE_SCHEMA_INVALID'
        return None

    def execute(self, request: dict[str, Any], action: dict[str, Any]) -> dict[str, Any]:
        request_digest = digest_data(request)
        if not self.verification.get('overall_ok', False):
            refusal = make_refusal(
                code=self.verification.get('refusal_code', 'REFUSE_BUNDLE_VERIFICATION_FAILED'),
                layer='bundle',
                bundle_digest=self.verification.get('bundle_digest'),
                request_digest=request_digest,
                verification_result='NOT_VERIFIED',
                refusal_reasons=self.verification.get('refusal_reasons', ['REFUSE_BUNDLE_VERIFICATION_FAILED']),
                details={'source': 'externalized.execute'},
            )
            refusal['overall_ok'] = False
            return attach_witness_certificate(refusal)
        request_problem = self._validate_request(request)
        if request_problem is not None:
            refusal = make_refusal(code=request_problem, layer='runtime', bundle_digest=self.bundle_digest, request_digest=request_digest, refusal_reasons=[request_problem], details={'source': 'externalized.execute'})
            refusal['overall_ok'] = False
            return attach_witness_certificate(refusal)
        if not isinstance(action, dict) or sorted(action.keys()) != ['action_id', 'kind', 'payload']:
            refusal = make_refusal(code='REFUSE_ACTION_DESCRIPTOR_INVALID', layer='action', bundle_digest=self.bundle_digest, request_digest=request_digest, action_digest=None, refusal_reasons=['REFUSE_ACTION_DESCRIPTOR_INVALID'], details={'source': 'externalized.execute'})
            refusal['overall_ok'] = False
            return attach_witness_certificate(refusal)
        result = evaluate_request(self.profile, request)
        if result.verdict != 'ALLOW':
            refusal = make_refusal(code='REFUSE_VERDICT_NOT_ALLOW', layer='action', bundle_digest=self.bundle_digest, request_digest=request_digest, action_digest=digest_data(action), refusal_reasons=['REFUSE_VERDICT_NOT_ALLOW'], details={'source': 'externalized.execute', 'decisive_rule_ids': result.decisive_rule_ids, 'matched': result.matched, 'failed': result.failed, 'fail_closed': result.fail_closed})
            refusal['overall_ok'] = False
            return attach_witness_certificate(refusal)
        return attach_witness_certificate(make_permit(bundle_digest=self.bundle_digest, request_digest=request_digest, action=action, decisive_rule_ids=result.decisive_rule_ids))


def replay_verdict_externalized(bundle_dir: str | Path, request: dict[str, Any]) -> dict[str, Any]:
    runner = ExternalizedRunner(bundle_dir)
    request_digest = digest_data(request)
    if not runner.verification.get('overall_ok', False):
        refusal = make_refusal(
            code=runner.verification.get('refusal_code', 'REFUSE_BUNDLE_VERIFICATION_FAILED'),
            layer='bundle',
            bundle_digest=runner.verification.get('bundle_digest'),
            request_digest=request_digest,
            verification_result='NOT_VERIFIED',
            refusal_reasons=runner.verification.get('refusal_reasons', ['REFUSE_BUNDLE_VERIFICATION_FAILED']),
            details={'source': 'externalized.replay'},
        )
        refusal['overall_ok'] = False
        return attach_witness_certificate(refusal)
    request_problem = runner._validate_request(request)
    if requestProblem := request_problem:
        refusal = make_refusal(code=requestProblem, layer='runtime', bundle_digest=runner.bundle_digest, request_digest=request_digest, refusal_reasons=[requestProblem], details={'source': 'externalized.replay'})
        refusal['overall_ok'] = False
        return attach_witness_certificate(refusal)
    result = evaluate_request(runner.profile, request)
    if result.verdict != 'ALLOW':
        refusal = make_refusal(code='REFUSE_VERDICT_NOT_ALLOW', layer='runtime', bundle_digest=runner.bundle_digest, request_digest=request_digest, refusal_reasons=['REFUSE_VERDICT_NOT_ALLOW'], details={'source': 'externalized.replay', 'decisive_rule_ids': result.decisive_rule_ids, 'matched': result.matched, 'failed': result.failed, 'fail_closed': result.fail_closed})
        refusal['overall_ok'] = False
        return attach_witness_certificate(refusal)
    return attach_witness_certificate({
        'primitive_identity': 'BCF Primitive One',
        'overall_ok': True,
        'bundle_digest': runner.bundle_digest,
        'request_digest': request_digest,
        'verdict': 'ALLOW',
        'replay_binding': {'primitive_identity': 'BCF Primitive One', 'bundle_digest': runner.bundle_digest, 'request_digest': request_digest},
        'decisive_rule_ids': list(result.decisive_rule_ids),
        'matched': list(result.matched),
        'failed': list(result.failed),
        'fail_closed': result.fail_closed,
    })
